File: apps/backend/src/schemas/customer_preferences.py
"""
Customer Preferences Schemas
============================

Pydantic schemas for the unified customer preferences capture system.
Handles chef requests, allergens, and special instructions.

Related Tables:
- core.bookings (stores preferences per booking)
- core.common_allergens (reference table)

Related Services:
- services/customer_preferences_service.py
"""

from datetime import datetime
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

# =====================================================
# Common Allergens Reference
# =====================================================


class CommonAllergenInfo(BaseModel):
    """A common allergen from the reference table."""

    code: str = Field(
        ..., description="Allergen code: shellfish, fish, soy, gluten, etc."
    )
    display_name: str = Field(..., description="Human-readable name with examples")
    icon: Optional[str] = Field(None, description="Emoji icon for UI display")
    chef_action: Optional[str] = Field(
        None, description="Cooking instructions for chef"
    )
    display_order: int = Field(0, description="Order in UI list")


class CommonAllergensListResponse(BaseModel):
    """Response containing all available common allergens."""

    allergens: List[CommonAllergenInfo]
    total: int


# =====================================================
# Chef Request Types
# =====================================================

ChefRequestSource = Literal[
    "online_form",  # Customer selected on booking form
    "phone",  # Customer mentioned during phone call
    "text",  # Customer mentioned in text/SMS
    "ai_chat",  # AI detected chef name in conversation
    "email",  # Customer mentioned in email
    "returning_customer",  # System prompt for returning customer
    "walk_in",  # In-person at event or office
]

AllergenConfirmMethod = Literal[
    "online",  # Confirmed via online booking form
    "sms",  # Confirmed via SMS reply
    "phone",  # Confirmed during phone call
    "email",  # Confirmed via email
    "in_person",  # Confirmed in person
]


# =====================================================
# Customer Preferences (Combined DTO)
# =====================================================


class ChefRequestInfo(BaseModel):
    """Chef request details."""

    chef_was_requested: bool = Field(
        False, description="True if customer requested a specific chef"
    )
    requested_chef_id: Optional[UUID] = Field(
        None, description="UUID of requested chef"
    )
    requested_chef_name: Optional[str] = Field(
        None, description="Name of requested chef (for display)"
    )
    chef_request_source: Optional[ChefRequestSource] = Field(
        None, description="How request was captured"
    )


class AllergenInfo(BaseModel):
    """Allergen and dietary information."""

    common_allergens: List[str] = Field(
        default_factory=list,
        description="List of allergen codes: shellfish, soy, eggs, dairy, sesame, nuts, gluten, fish, msg",
    )
    allergen_disclosure: Optional[str] = Field(
        None, max_length=1000, description="Free text for other allergens/dietary needs"
    )
    allergen_confirmed: bool = Field(
        False, description="True if info was confirmed with customer"
    )
    allergen_confirmed_method: Optional[AllergenConfirmMethod] = Field(None)
    allergen_confirmed_at: Optional[datetime] = Field(None)
    allergen_confirmed_by_id: Optional[UUID] = Field(
        None, description="Staff who confirmed"
    )
    allergen_confirmed_by_name: Optional[str] = Field(
        None, description="Staff name (for display)"
    )


class CustomerPreferencesResponse(BaseModel):
    """Full customer preferences for a booking (read)."""

    booking_id: UUID

    # Chef Request
    chef_request: ChefRequestInfo

    # Allergens
    allergens: AllergenInfo

    # Special Instructions (existing field)
    special_requests: Optional[str] = Field(
        None, description="General special instructions"
    )

    # Metadata
    last_updated: Optional[datetime] = None


class CustomerPreferencesUpdate(BaseModel):
    """Update customer preferences for a booking."""

    # Chef Request (optional section)
    chef_was_requested: Optional[bool] = None
    requested_chef_id: Optional[UUID] = None
    chef_request_source: Optional[ChefRequestSource] = None

    # Allergens (optional section)
    common_allergens: Optional[List[str]] = None
    allergen_disclosure: Optional[str] = Field(None, max_length=1000)
    allergen_confirmed: Optional[bool] = None
    allergen_confirmed_method: Optional[AllergenConfirmMethod] = None

    # Special Instructions
    special_requests: Optional[str] = Field(None, max_length=2000)


# =====================================================
# Chef Prep Summary (for Chef App)
# =====================================================


class AllergenPrepNote(BaseModel):
    """Single allergen with cooking action for chef."""

    allergen: str = Field(..., description="Allergen code")
    display_name: str = Field(..., description="Human-readable name")
    icon: str = Field("⚠️", description="Emoji icon")
    chef_action: str = Field(..., description="What chef should do")


class ChefPrepAllergensResponse(BaseModel):
    """Allergen summary for chef prep view."""

    booking_id: UUID
    customer_name: str
    event_datetime: datetime

    # Allergens
    has_allergens: bool = Field(False)
    allergen_notes: List[AllergenPrepNote] = Field(default_factory=list)
    other_dietary: Optional[str] = Field(None, description="Free text dietary notes")

    # Confirmation status
    allergen_confirmed: bool = Field(False)
    allergen_confirmed_method: Optional[str] = None

    # Special requests
    special_requests: Optional[str] = None


# =====================================================
# Bonus Calculation Response
# =====================================================


class ChefRequestBonusInfo(BaseModel):
    """Information about chef request bonus for a booking."""

    booking_id: UUID
    chef_was_requested: bool = False
    requested_chef_id: Optional[UUID] = None
    requested_chef_name: Optional[str] = None
    assigned_chef_id: Optional[UUID] = None
    assigned_chef_name: Optional[str] = None

    # Bonus calculation
    bonus_eligible: bool = Field(
        False, description="True if requested chef = assigned chef"
    )
    bonus_percentage: int = Field(
        10, description="Bonus percentage from SSoT (default 10%)"
    )
    base_order_cents: int = Field(
        0, description="Base order value (adults × rate + kids × rate)"
    )
    bonus_amount_cents: int = Field(0, description="Calculated bonus amount")

File: apps/admin/src/types/customer-preferences.ts
/**
 * Customer Preferences Types
 * ==========================
 *
 * Types for the unified customer preferences capture system:
 * - Chef request tracking (for 10% bonus)
 * - Allergen disclosure
 *
 * These match the backend schemas in:
 * apps/backend/src/schemas/customer_preferences.py
 *
 * Database Tables:
 * - core.bookings (6 new columns from migration 022)
 * - core.common_allergens (9 standard allergens)
 *
 * API Endpoints:
 * - GET /api/v1/allergens
 * - GET /api/v1/bookings/{id}/preferences
 * - PUT /api/v1/bookings/{id}/preferences
 * - GET /api/v1/bookings/{id}/prep-allergens
 * - GET /api/v1/bookings/{id}/chef-bonus
 */

/**
 * Common allergen from the reference table
 * Uses 'code' as primary identifier (matches DB column)
 */
export interface CommonAllergen {
  code: string;
  display_name: string;
  icon: string;
}

/**
 * Chef request source options
 * Tracks HOW the customer requested a specific chef
 */
export type ChefRequestSource =
  | 'booking_form' // Customer selected on website booking form
  | 'sms' // Customer requested via SMS
  | 'phone_call' // Customer mentioned during phone call
  | 'email' // Customer requested via email
  | 'social_dm' // Customer requested via social media DM
  | 'ai_chat' // Customer mentioned to AI assistant
  | 'repeat_customer' // Auto-detected from previous bookings
  | 'admin_manual'; // Admin manually marked as requested

/**
 * Chef request info for a booking
 */
export interface ChefRequestInfo {
  chef_was_requested: boolean;
  requested_chef_id: string | null;
  requested_chef_name: string | null;
  chef_request_source: ChefRequestSource | null;
}

/**
 * Allergen info for a booking
 */
export interface AllergenInfo {
  allergen_disclosure: string | null;
  common_allergens: string[]; // Array of allergen codes
  allergen_confirmed: boolean;
}

/**
 * Full customer preferences response (combines chef request + allergens)
 */
export interface CustomerPreferences {
  booking_id: string;
  chef_request: ChefRequestInfo;
  allergens: AllergenInfo;
}

/**
 * Update request for customer preferences
 * All fields are optional - only send what you want to update
 */
export interface CustomerPreferencesUpdate {
  // Chef request fields
  chef_was_requested?: boolean;
  requested_chef_id?: string | null;
  chef_request_source?: ChefRequestSource;

  // Allergen fields
  allergen_disclosure?: string;
  common_allergens?: string[]; // Array of allergen codes
  allergen_confirmed?: boolean;
}

/**
 * Chef prep view of allergens
 * Simplified view for chef mobile app / prep checklist
 */
export interface ChefPrepAllergens {
  booking_id: string;
  has_allergens: boolean;
  common_allergens: CommonAllergen[]; // Full allergen objects with display info
  additional_disclosure: string | null;
  confirmed: boolean;
  cooking_actions: CookingAction[];
}

/**
 * Cooking action for allergen accommodation
 */
export interface CookingAction {
  allergen_code: string;
  allergen_name: string;
  action: string; // e.g., "Cook shrimp LAST on separate section of grill"
}

/**
 * Chef bonus calculation result
 */
export interface ChefBonusInfo {
  booking_id: string;
  chef_was_requested: boolean;
  requested_chef_id: string | null;
  assigned_chef_id: string | null;
  bonus_applies: boolean; // True if requested == assigned
  bonus_pct: number; // e.g., 10
  base_earnings_cents: number;
  bonus_cents: number;
  message: string;
}

/**
 * Request source display info
 */
export const CHEF_REQUEST_SOURCE_LABELS: Record<ChefRequestSource, string> = {
  booking_form: 'Booking Form',
  sms: 'SMS',
  phone_call: 'Phone Call',
  email: 'Email',
  social_dm: 'Social Media DM',
  ai_chat: 'AI Chat',
  repeat_customer: 'Repeat Customer',
  admin_manual: 'Admin Manual',
};

/**
 * Request source icons (for UI)
 */
export const CHEF_REQUEST_SOURCE_ICONS: Record<ChefRequestSource, string> = {
  booking_form: '🌐',
  sms: '💬',
  phone_call: '📞',
  email: '📧',
  social_dm: '📱',
  ai_chat: '🤖',
  repeat_customer: '🔄',
  admin_manual: '✏️',
};
